ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/authorization/ApiServiceAuthorization.scala

```diff
@@ -0,0 +1,111 @@
+import io.grpc.{Context, StatusRuntimeException}
+import org.slf4j.{Logger, LoggerFactory}
+
+sealed abstract class Authorization
+case object Authorized extends Authorization
+final case class NotAuthorized(reason: String = "You are not authorized to use this resource")
+    extends Authorization
+
+/** A simple helper that allows services to use authorization claims
+  * that have been stored by [[AuthorizationInterceptor]].
+  */
+object ApiServiceAuthorization {
+
+  protected val logger: Logger = LoggerFactory.getLogger(ApiServiceAuthorization.getClass)
+
+  private[this] def getStackTrace(t: Throwable): String = {
+    import java.io.PrintWriter
+    import java.io.StringWriter
+    val sw = new StringWriter
+    val pw = new PrintWriter(sw, true)
+    t.printStackTrace(pw)
+    pw.flush()
+    sw.flush()
+    sw.toString
+  }
+
+  /** Checks whether the given claims give access to the request.
+    *
+    * @param check Given a set of [[Claims]], returns whether the current method is authorized.
+    *              Typically, the caller compares the request parameters with the given claims.
+    *
+    * @return   A [[StatusRuntimeException]] with code `PERMISSION_DENIED` if the request is not authorized,
+    *           or Unit if the request is authorized.
+    */
+  def requireClaims(check: Claims => Authorization): Either[StatusRuntimeException, Unit] = {
+    val claims = AuthorizationInterceptor.contextKeyClaim.get()
+    if (claims == null) {
+      // The current context does not contain any claims.
+      // Most likely causes:
+      // - The API server does not use the [[AuthorizationInterceptor]] which is responsible for creating the context
+      // - This function is called from a thread different from the one used to handle the gRPC call,
+      //   such as during a `Future.map`. See also [[ApiServiceAuthorization.withContext]].
+      logger.error(s"No context key at: ${getStackTrace(new Exception)}")
+      Left(internal("Context.key.get returned no valid Claims."))
+    } else {
+      check(claims) match {
+        case Authorized =>
+          Right(())
+        case NotAuthorized(reason) =>
+          // Be careful with logging errors in order to not leak secrets through log messages
+          // logger.warn(s"Rejecting authorization with claims $claims")
+          Left(permissionDenied(reason))
+      }
+    }
+  }
+
+  /** Executes the given block with the given gRPC context as the current context.
+    *
+    * Usage:
+    * ```
+    * // Capture the gRPC context in a gRPC thread
+    * val ctx = Context.current
+    *
+    * // Function bar uses ApiServiceAuthorization.requireClaims,
+    * // but is called from a non-gRPC thread.
+    * // Wrap it in withContext, so that the gRPC context is available in bar()
+    * val asyncResult: Future[T] = foo()
+    *   .flatMap(ApiServiceAuthorization.withContext(ctx)(bar))
+    * ```
+    */
+  def withContext[T](context: Context)(block: => T): T = {
```

Comment:
I don't see usages of this method in this PR, am I missing something?

Reply:
No usages, but a useful method when dealing with gRPC contexts. We might need it later for implementing canceling requests after claims expiration, so I did not want to delete it.